src/nvim/terminal.c

```diff
@@ -353,8 +353,14 @@ void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
   invalidate_terminal(term, -1, -1);
 }
 
-void terminal_enter(Terminal *term, bool process_deferred)
+void terminal_enter(bool process_deferred)
 {
+  Terminal *term = curbuf->terminal;
+  assert(term && "should only be called when curbuf has a terminal");
```

Comment:
> I see that the terminal also has a reference to the buffer. Maybe this is a sign that most the function should actually take the buffer as argument.

Yes, a `Terminal` is always tied to a `buf_T`. And since `buf_T->terminal` is used to distinguish normal buffers from terminal buffers, I guess it makes sense to refactor most functions in this module to receive buffers, thus avoiding the need for cyclic references(In another PR, of course)